hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java

```diff
@@ -199,29 +203,68 @@ protected void doCommit(TableMetadata base, TableMetadata metadata) {
       setHmsTableParameters(newMetadataLocation, tbl, metadata.properties(), removedProps, hiveEngineEnabled);
 
       persistTable(tbl, updateHiveTable);
-      threw = false;
     } catch (org.apache.hadoop.hive.metastore.api.AlreadyExistsException e) {
+      canCleanupMetadata = true;
       throw new AlreadyExistsException("Table already exists: %s.%s", database, tableName);
 
     } catch (TException | UnknownHostException e) {
```

Comment:
I wonder whether this place is too generic. Apart from handling exceptions in alter_table (i.e. the actual commit), we also catch any exceptions during loading and locking. We don't necessarily have to do these checks in cases where we did not attempt to commit a new version.

Is it a crazy idea to only add this check in `persistTable`, the one that does the actual commit?

Reply:
Then we could handle all exception types there and don't worry about getting only `TException`.

Reply:
I think it may be sufficient to do this:

```
try {
  persistTable(tbl, updateHiveTable);
} catch (Exception commitException) {
  try {
    canCleanupMetadata = !isCommitted(newMetadataLocation);
  } catch (Exception checkException) {
    LOG.error("Cannot determine if commit was successful", checkException);
    throw new CommitStateUnknownException(commitException);
  }
}
```

```
private boolean isCommitted(String newMetadataLocation) {
  TableMetadata metadata = refresh();
  String metadataLocation = metadata.metadataFileLocation();
  return metadataLocation.equals(newMetadataLocation) ||
      metadata.previousFiles().stream().anyMatch(log -> log.file().equals(newMetadataLocation));
}
```

Then we don't need extra checks in catch clauses.

Reply:
Switched it around, we have to set canCleanupMetadata to false before calling persistTable which looks a little weird but I think overall it's a little cleaner

Reply:
I think this is still not very clear, but If other folks can read it let's keep it.

Reply:
And this apparently violates one of our checkstyle rules for Try nesting :(

Reply:
ugh and it fails because we need to rethrow the exception if canCleanupMetadata is false